// humanize/specs/dateData.js
module.exports = function() {

  // all in America/Los_Angeles timezone

  var timestamps = {
    // 04/16/1986 18:23:47
    '514088627': {
      d: '16',
      D: 'Wed',
      j: '16',
      l: 'Wednesday',
      N: '3',
      S: 'th',
      w: '3',
      z: '105',
      W: '16',
      F: 'April',
      m: '04',
      M: 'Apr',
      n: '4',
      t: '30',
      L: '0',
      o: '1986',
      Y: '1986',
      y: '86',
      a: 'pm',
      A: 'PM',
      B: '141',
      g: '6',
      G: '18',
      h: '06',
      H: '18',
      i: '23',
      s: '47',
      u: '000000',
      // I: '0',
      O: '-0800',
      P: '-08:00',
      Z: '-28800',
      c: '1986-04-16T18:23:47-08:00',
      r: 'Wed, 16 Apr 1986 18:23:47 -0800',
      U: '514088627'
    },

    // 12/25/1999 01:09:03
    '946112943': {
      d: '25',
      D: 'Sat',
      j: '25',
      l: 'Saturday',
      N: '6',
      S: 'th',
      w: '6',
      z: '358',
      W: '51',
      F: 'December',
      m: '12',
      M: 'Dec',
      n: '12',
      t: '31',
      L: '0',
      o: '1999',
      Y: '1999',
      y: '99',
      a: 'am',
      A: 'AM',
      B: '422',
      g: '1',
      G: '1',
      h: '01',
      H: '01',
      i: '09',
      s: '03',
      u: '000000',
      // I: '0',
      O: '-0800',
      P: '-08:00',
      Z: '-28800',
      c: '1999-12-25T01:09:03-08:00',
      r: 'Sat, 25 Dec 1999 01:09:03 -0800',
      U: '946112943'
    },

    // 07/10/2011 12:23:58
    '1310325838': {
      d: '10',
      D: 'Sun',
      j: '10',
      l: 'Sunday',
      N: '7',
      S: 'th',
      w: '0',
      z: '190',
      W: '27',
      F: 'July',
      m: '07',
      M: 'Jul',
      n: '7',
      t: '31',
      L: '0',
      o: '2011',
      Y: '2011',
      y: '11',
      a: 'pm',
      A: 'PM',
      B: '849',
      g: '12',
      G: '12',
      h: '12',
      H: '12',
      i: '23',
      s: '58',
      u: '000000',
      // I: '1',
      O: '-0700',
      P: '-07:00',
      Z: '-25200',
      c: '2011-07-10T12:23:58-07:00',
      r: 'Sun, 10 Jul 2011 12:23:58 -0700',
      U: '1310325838'
    },

    // 10/05/2015 00:00:00
    '1444028400': {
      d: '05',
      D: 'Mon',
      j: '5',
      l: 'Monday',
      N: '1',
      S: 'th',
      w: '1',
      z: '277',
      W: '41',
      F: 'October',
      m: '10',
      M: 'Oct',
      n: '10',
      t: '31',
      L: '0',
      o: '2015',
      Y: '2015',
      y: '15',
      a: 'am',
      A: 'AM',
      B: '333',
      g: '12',
      G: '0',
      h: '12',
      H: '00',
      i: '00',
      s: '00',
      u: '000000',
      // I: '1',
      O: '-0700',
      P: '-07:00',
      Z: '-25200',
      c: '2015-10-05T00:00:00-07:00',
      r: 'Mon, 05 Oct 2015 00:00:00 -0700',
      U: '1444028400'
    },

    // 03/31/2000 13:25:14
    '954537914': {
      d: '31',
      D: 'Fri',
      j: '31',
      l: 'Friday',
      N: '5',
      S: 'st',
      w: '5',
      z: '90',
      W: '13',
      F: 'March',
      m: '03',
      M: 'Mar',
      n: '3',
      t: '31',
      L: '1',
      o: '2000',
      Y: '2000',
      y: '00',
      a: 'pm',
      A: 'PM',
      B: '934',
      g: '1',
      G: '13',
      h: '01',
      H: '13',
      i: '25',
      s: '14',
      u: '000000',
      // I: '0',
      O: '-0800',
      P: '-08:00',
      Z: '-28800',
      c: '2000-03-31T13:25:14-08:00',
      r: 'Fri, 31 Mar 2000 13:25:14 -0800',
      U: '954537914'
    },

    // 01/02/3456 12:34:56
    '46893760496': {
      d: '02',
      D: 'Wed',
      j: '2',
      l: 'Wednesday',
      N: '3',
      S: 'nd',
      w: '3',
      z: '1',
      W: '01',
      F: 'January',
      m: '01',
      M: 'Jan',
      n: '1',
      t: '31',
      L: '1',
      o: '3456',
      Y: '3456',
      y: '56',
      a: 'pm',
      A: 'PM',
      B: '899',
      g: '12',
      G: '12',
      h: '12',
      H: '12',
      i: '34',
      s: '56',
      u: '000000',
      // I: '0',
      O: '-0800',
      P: '-08:00',
      Z: '-28800',
      c: '3456-01-02T12:34:56-08:00',
      r: 'Wed, 02 Jan 3456 12:34:56 -0800',
      U: '46893760496'
    },

    // 01/11/1840 11:56:14
    '-4101509026': {
      d: '11',
      D: 'Sat',
      j: '11',
      l: 'Saturday',
      N: '6',
      S: 'th',
      w: '6',
      z: '10',
      W: '02',
      F: 'January',
      m: '01',
      M: 'Jan',
      n: '1',
      t: '31',
      L: '1',
      o: '1840',
      Y: '1840',
      y: '40',
      a: 'am',
      A: 'AM',
      B: '873',
      g: '11',
      G: '11',
      h: '11',
      H: '11',
      i: '56',
      s: '14',
      u: '000000',
      // I: '0',
      O: '-0800',
      P: '-08:00',
      Z: '-28800',
      c: '1840-01-11T11:56:14-08:00',
      r: 'Sat, 11 Jan 1840 11:56:14 -0800',
      U: '-4101509026'
    }
  };

  return {
    timestamps: timestamps
  };

};
